### ISSUE TITLE
[`unnecessary_parenthesis`] false negative in conditional operator

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#4996]
When doing this:

```dart
final email = (user.email != null) ? (user.email!) : '';
```

This is a dumb example, but my point is that in other ternary expressions that I've used, I saw this.

I _think_ that the `(user.email!)` expression should trigger this as well. I'm not sure because this might be intentional to leave ternary expressions with possible parentheses around all possible results and the test, but it seems strange to me that it doesn't warn me when that is not strictly necessary.

This is a false negative that occurs in any of the three places of a ternary operator.

_Originally posted by @FMorschel in https://github.com/dart-lang/linter/issues/4980#issuecomment-2152506202_

**Comment by lrhn** [comment]
It does appear there is an exception for the brances of conditional expressions.
Maybe that exception doesn't need to apply to all expressions. A `primary + selectorChain` may not need parentheses. Maybe suffix/prefix increment operators can be included too, so `c ? foo.bar.baz[42].qux++ : x++` won't allow parentheses.  

I usually only want parentheses when the syntax contains whitespace, so something with binary operators or prefix operators that are words, fx `await`. Prefix/suffix operators are flush against the expression, and selector chains only contain spaces inside argument lists, type argument lists and string literals, which are already delimited. Could consider not exempting those from the unnecessary parentheses lint.


(Pet-quixotic-peeve: The `?`/`:` is the conditional operator/expression. It's sometimes called ["the ternary conditional operator" or "ternary if"](https://en.wikipedia.org/wiki/Ternary_conditional_operator#:~:text=It%20is%20commonly%20referred%20to,a%20then%20b%20otherwise%20c%22.), but it's imprecise to just say "the ternary operator". It's *a* ternary operator, but `[]=` is also a ternary operator in Dart - it takes three operands. This is likely a lost cause, but by golly, I'm going to keep trying!)

**Comment by srawlins** [comment]
Similar to https://github.com/dart-lang/linter/issues/4980, this is a case where the rule does not report the parentheses, because they can improve readability. See the test case, `test_notEqualInside_returnExpression` (`return (1 != 1);`), which has parens that are not reported.

**Comment by FMorschel** [comment]
Yes, I do agree that they can make readability better. I just disagree that the following has no warnings:

```dart
(list.isEmpty) ? (false) : (false);
```

Neither of the three has spaces or would be more readable because of the parentheses.

**Comment by srawlins** [comment]
Ah I missed the `user.email!` case.

I just tested at head, and one lint diagnostic is reported for:

```dart
class U {
  String? email = '';
}
var u = U();
var x = 1 == 2 ? (u.email!) : false;
```

but none are reported for:

```
var list = [];
var x = (list.isEmpty) ? true : false;
```

I agree something should be reported for the second case.

**Comment by FMorschel** [comment]
I've downloaded the latest sdk minutes ago (from [here](https://gsdview.appspot.com/dart-archive/channels/main/raw/latest/sdk/)) and I got no lints in:

```dart
class U {
  String? email = '';
}
var u = U();
var list = [];
var x = 1 == 2 ? (u.email!) : (list.isEmpty);
```

I'm not sure what you got. I do have `unnecessary_parenthesis` set to `true`.

**Comment by FMorschel** [comment]
I've opened https://github.com/dart-lang/linter/issues/4996

**Comment by FMorschel** [comment]
I've opened drafts for fixing the new triggers for this lint. As soon as the bots are run again (made some more changes for the last review) I can go through them again and check if anything else changed to complete them.

- https://github.com/flutter/flutter/pull/156928
- https://github.com/flutter/packages/pull/7881

**Comment by pq** [comment]
Thanks!

**Comment by FMorschel** [comment]
I've reviewed everything. As soon as I get the resolved state on the last review I'll ask for the PRs to be reviewed and merged, and then we'll be good to go. Nothing has changed since the last run (which I was expecting).

**Comment by nate-thegrate** [comment]
> (Pet-quixotic-peeve: The `?`/`:` is the conditional operator/expression. It's sometimes called ["the ternary conditional operator" or "ternary if"](https://en.wikipedia.org/wiki/Ternary_conditional_operator#:~:text=It%20is%20commonly%20referred%20to,a%20then%20b%20otherwise%20c%22.), but it's imprecise to just say "the ternary operator". It's _a_ ternary operator, but `[]=` is also a ternary operator in Dart - it takes three operands. This is likely a lost cause, but by golly, I'm going to keep trying!)

Arrived here after reading through https://github.com/flutter/flutter/pull/156928. Proud to say I support this cause 🫡

**Comment by FMorschel** [comment]
As an update in case anyone is not subscribed to the PRs, they have been merged. We're now only missing the long list of internal code triggers for this change.
